File: docs-source/docs/modules/cqrs/examples/shopping-cart-service-scala/src/main/scala/sample/shoppingcart/ShoppingCartServer.scala
package sample.shoppingcart

import scala.concurrent.duration._
import scala.concurrent.ExecutionContext
import scala.concurrent.Future
import scala.util.Failure
import scala.util.Success

import akka.actor.typed.ActorSystem
import akka.grpc.scaladsl.ServerReflection
import akka.grpc.scaladsl.ServiceHandler
import akka.http.scaladsl.Http
import akka.http.scaladsl.model.HttpRequest
import akka.http.scaladsl.model.HttpResponse

class ShoppingCartServer(port: Int, system: ActorSystem[_]) {
  private implicit val sys: ActorSystem[_] = system
  private implicit val ec: ExecutionContext = system.executionContext

  def start(): Unit = {

    val service: HttpRequest => Future[HttpResponse] =
      ServiceHandler.concatOrNotFound(
        proto.ShoppingCartServiceHandler.partial(new ShoppingCartServiceImpl()),
        // ServerReflection enabled to support grpcurl without import-path and proto parameters
        ServerReflection.partial(List(proto.ShoppingCartService)))

    val bound =
      Http().newServerAt(interface = "127.0.0.1", port).bind(service).map(_.addToCoordinatedShutdown(3.seconds))

    bound.onComplete {
      case Success(binding) =>
        val address = binding.localAddress
        system.log.info("Shopping online at gRPC server {}:{}", address.getHostString, address.getPort)
      case Failure(ex) =>
        system.log.error("Failed to bind gRPC endpoint, terminating system", ex)
        system.terminate()
    }
  }

}

File: docs-source/docs/modules/microservices-tutorial/examples/01-shopping-cart-service-java/src/main/java/shopping/cart/ShoppingCartServer.java
package shopping.cart;

import akka.actor.typed.ActorSystem;
import akka.grpc.ServiceDescription;
import akka.grpc.javadsl.ServiceHandler;
import akka.grpc.javadsl.ServerReflection;
import akka.http.javadsl.Http;
import akka.http.javadsl.ServerBinding;
import akka.http.javadsl.model.HttpRequest;
import akka.http.javadsl.model.HttpResponse;
import akka.japi.Function;
import shopping.cart.proto.ShoppingCartServiceHandlerFactory;
import shopping.cart.proto.ShoppingCartService;

import java.net.InetSocketAddress;
import java.time.Duration;
import java.util.Arrays;
import java.util.Collections;
import java.util.concurrent.CompletionStage;

public final class ShoppingCartServer {

    private ShoppingCartServer() { }

    static void start(String host, int port, ActorSystem<?> system, ItemPopularityRepository itemPopularityRepository) {
        ShoppingCartServiceImpl impl = new ShoppingCartServiceImpl(system, itemPopularityRepository);
        @SuppressWarnings("unchecked")
        Function<HttpRequest, CompletionStage<HttpResponse>> service =
                ServiceHandler.concatOrNotFound(
                    ShoppingCartServiceHandlerFactory.create(impl, system),
                    // ServerReflection enabled to support grpcurl without import-path and proto parameters
                    ServerReflection.create(Collections.singletonList(ShoppingCartService.description), system));

        CompletionStage<ServerBinding> bound = Http.get(system).newServerAt(host, port)
                .bind(service::apply);

        bound.whenComplete((binding, ex) -> {
            if (binding != null) {
                binding.addToCoordinatedShutdown(Duration.ofSeconds(3), system);
                InetSocketAddress address = binding.localAddress();
                system.log().info("Shopping online at gRPC server {}:{}", address.getHostString(), address.getPort());
            } else {
                system.log().error("Failed to bind gRPC endpoint, terminating system", ex);
                system.terminate();
            }
        });
    }
}
